in_toto/gpg/common.py

```diff
@@ -127,62 +134,363 @@ def parse_pubkey_payload(data):
+
+      else:
+        log.info("Ignoring gpg key packet '{}', we only handle packets of "
+            "types '{}' (see RFC4880 4.3. Packet Tags).".format(packet_type,
+            [PACKET_TYPE_PRIMARY_KEY, PACKET_TYPE_USER_ID,
+            PACKET_TYPE_USER_ATTR, PACKET_TYPE_SUB_KEY,
+            PACKET_TYPE_SIGNATURE]))
+
+    except Exception as e:
+      raise PacketParsingError("Invalid public key data at position {}: {}."
+          .format(position, e))
+
+    # Go to next packet
+    position += packet_length
+
+  # Parsing is done. Now enrich the master key with certificate info and
+  # verify the subkeys bindings.
+  master_key = _assign_certified_key_info(key_bundle)
+  verified_subkeys = _get_verified_subkeys(key_bundle)
+
+  return master_key, verified_subkeys
+
+
+def _assign_certified_key_info(bundle):
+  """
+  <Purpose>
+    Helper function to verify User ID certificates corresponding to a gpg
+    master key, in order to enrich the master key with additional information
+    (e.g. expiration dates). The enriched master key is returned.
+
+    TODO: Do extract the information and address ambiguity (see inline TODO
+    below).
+
+    NOTE: Currently we only consider User ID certificates. We can do the same
+    for User Attribute certificates by iterating over
+    bundle[PACKET_TYPE_USER_ATTR] instead of bundle[PACKET_TYPE_USER_ID], and
+    replacing the signed_content constant '\xb4'  with '\xd1' (see RFC4880
+    section 5.2.4. paragraph 4).
+
+  <Arguments>
+    bundle:
+          GPG key bundle as parsed in parse_pubkey_bundle().
+
+  <Exceptions>
+    None.
+
+  <Side Effects>
+    None.
+
+  <Returns>
+    A public key in the format in_toto.gpg.formats.PUBKEY_SCHEMA.
 
   """
-  master_public_key = None
-  sub_public_keys = {}
+  # Create handler shortcut
+  handler = SIGNATURE_HANDLERS[bundle[PACKET_TYPE_PRIMARY_KEY]["key"]["type"]]
+
+  # Verify User ID signatures to gather information about primary key
+  # (see Notes about certification signatures in RFC 4880 5.2.3.3.)
+  for user_id_packet, packet_data in bundle[PACKET_TYPE_USER_ID].items():
+    try:
+      # Construct signed content (see RFC4880 section 5.2.4. paragraph 4)
+      signed_content = (bundle[PACKET_TYPE_PRIMARY_KEY]["packet"] +
+          b"\xb4\x00\x00\x00" + user_id_packet[1:])
+
+    except Exception as e:
+      log.info(e)
+      continue
+
+    for signature_packet in packet_data["signatures"]:
+      try:
+        signature = parse_signature_packet(signature_packet,
+            supported_hash_algorithms={SHA1, SHA256, SHA512},
+            supported_signature_types=SIGNATURE_TYPE_CERTIFICATES,
+            include_info=True)
+        # gpg_verify_signature requires a "keyid" even if it is short.
+        # (see parse_signature_packet for more information about keyids)
+        signature["keyid"] = signature["keyid"] or signature["short_keyid"]
+
+      except Exception as e:
+        log.info(e)
+        continue
```

Comment:
same here

Reply:
Here we shouldn't because `parse_signature_packet` might fail legitimately, because we want to continue if parse_signature_packet fails because we're giving it e.g. a revocation signature, which might happen. In a more recent commit I added a `TODO: Revise exception taxonomy`